src/vmm/src/devices/virtio/net/device.rs

```diff
@@ -141,37 +144,56 @@ impl RxBuffers {
         Ok(())
     }
 
-    /// Returns the number of available `iovec` objects.
+    /// Returns the total size of available space in the buffer.
     #[inline(always)]
-    fn len(&self) -> usize {
+    fn capacity(&self) -> usize {
         self.iovec.len()
     }
 
-    /// Returns `true` if there aren't any available `iovec` objects.
-    #[inline(always)]
-    fn is_empty(&self) -> bool {
-        self.len() == 0
-    }
-
     /// Mark the first `size` bytes of available memory as used.
     ///
     /// # Safety:
     ///
     /// * The `RxBuffers` should include at least one parsed `DescriptorChain`.
     /// * `size` needs to be smaller or equal to total length of the first `DescriptorChain` stored
     ///   in the `RxBuffers`.
-    unsafe fn mark_used(&mut self, size: u32) -> ParsedDescriptorChain {
-        // Since we were able to write a frame in guest memory, we should have at least one
-        // descriptor chain here. If not, we have a bug, so fail fast, since the device is
-        // fundamentally broken.
-        let mut parsed_dc = self.parsed_descriptors.pop_front().expect(
-            "net: internal bug. Mismatch between written frame size and available descriptors",
-        );
+    unsafe fn mark_used(&mut self, mut bytes_written: u32, rx_queue: &mut Queue) {
+        self.used_bytes = bytes_written;
+
+        let mut used_heads: u16 = 0;
+        let mut write_used = |head_index: u16, bytes_written: u32, rx_queue: &mut Queue| {
+            if let Err(err) = rx_queue.write_used_element(
+                (rx_queue.next_used + Wrapping(self.used_descriptors)).0,
+                head_index,
+                bytes_written,
+            ) {
+                error!(
+                    "net: Failed to add used descriptor {} of length {} to RX queue: {err}",
+                    head_index, bytes_written
+                );
+            }
+            used_heads += 1;
+            self.used_descriptors += 1;
+        };
+
+        for parsed_dc in self.parsed_descriptors.iter() {
+            if bytes_written <= parsed_dc.length {
+                write_used(parsed_dc.head_index, bytes_written, rx_queue);
+                break;
+            } else {
+                write_used(parsed_dc.head_index, parsed_dc.length, rx_queue);
+                bytes_written -= parsed_dc.length;
+            }
+        }
 
-        self.header_set_num_buffers(1);
-        self.iovec.drop_descriptor_chain(&parsed_dc);
-        parsed_dc.length = size;
-        parsed_dc
+        self.header_set_num_buffers(used_heads);
+        for _ in 0..used_heads {
+            let parsed_dc = self
+                .parsed_descriptors
+                .pop_front()
+                .expect("This should never happen if write to the buffer succeded.");
+            self.iovec.drop_descriptor_chain(&parsed_dc);
+        }
```

Comment:
Can we avoid the closure and second loop by doing

```rs
if bytes_written == 0 {
  return Ok(());
}

let mut used_heads = 0u16;
while let Some(parsed_dc) = self.parsed_descriptors.pop_front() {
  let used_bytes = bytes_written.min(parsed_dc.length);
  rx_queue.write_used_element(rx_queue.next_used + Wrapping(used_heads)).0, parsed_dc.head_index, used_bytes).unwrap(); // Safe because we know head_index isn't out of bounds
  bytes_written -= used_bytes;
  self.iovec.drop_descriptor_chain(&parsed_dc);
  used_heads += 1;

  if bytes_written == 0 {
    break;
  }
}

self.header_set_num_buffers(used_heads);
self.used_descriptors += used_heads;
```

?

Reply:
The second loop is only avoidable if we store the pointer to the first `iovec` before doing any `drop_descriptor_chain`. Otherwise `header_set_num_buffers` will set `used_heads` into incorrect place (as the chains we used for packet are already dropped).

Reply:
Removed the closure.

Reply:
Ah, missed that. Thanks!